appserver/payara-appserver-modules/microprofile/openapi/src/main/java/fish/payara/microprofile/openapi/impl/visitor/OpenApiWalker.java

```diff
@@ -128,13 +137,30 @@ public OpenApiWalker(OpenAPI api, Set<Class<?>> allowedClasses, Map<String, Set<
             return 1;
         });
         this.classes.addAll(allowedClasses);
+        addInnerClasses(); // must happen before generating the resource mapping
+        this.resourceMapping = generateResourceMapping(classes);
+    }
+
+    private void addInnerClasses() {
+        List<Class<?>> topLevelClasses = new ArrayList<>(classes);
+        for (Class<?> topLevelClass : topLevelClasses) {
+            addInnerClasses(topLevelClass);
+        }
+    }
+
+    private void addInnerClasses(Class<?> topLevelClass) {
+        if (topLevelClass != null) {
+            classes.addAll(Arrays.asList(topLevelClass.getDeclaredClasses()));
+            if (topLevelClass.getSuperclass() != Object.class) {
+                addInnerClasses(topLevelClass.getSuperclass());
+            }
+        }
     }
 
     @Override
     public void accept(ApiVisitor visitor) {
         // OpenAPI necessary annotations
-        processAnnotations(OpenAPIDefinition.class, visitor::visitOpenAPI, true);
-        processAnnotations(Schema.class, visitor::visitSchema, true);
```

Comment:
FYI: `Schema` was processed twice which was needed previously but now works when keeping the later processing of `Schema` annotation.